Add option to disable JWT-SVID profile on spire server

nikotih opened the pull request:
<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/main/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco
-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**
<!-- Please provide a description of the affected functionality -->
Spire server JWT-SVID

**Description of change**
<!-- Please provide a description of the change -->
New option to disable JWT-SVID in spire server. JWT keys are no longer managed. Server and Agent api endpoints related to JWT return precondition failed grpc code

Answers from the issue:
- returning FailedPrecondition grpc code in JWT related endpoints

- decided to go with disable_jwt so dont have to go through all unit tests setting enable_jwt to true (but let me know what you prefer)

- example calls:
```
time="2025-08-26T13:35:22Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=MintJWTSVID request_id=5f535de6-d5e5-466a-beb4-ceab4f32e88f service=svid.v1.SVID subsystem_name=api
time="2025-08-26T13:35:22Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=PrepareJWTAuthority request_id=a756696c-c986-4e10-8f41-7c10a769f9e3 service=localauthority.v1.LocalAuthority subsystem_name=api
time="2025-08-26T13:35:23Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=GetJWTAuthorityState request_id=4bdbcaf6-cfbd-4a6b-abe4-0982d2502f88 service=localauthority.v1.LocalAuthority subsystem_name=api
time="2025-08-26T13:35:23Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=ActivateJWTAuthority request_id=8e6191f2-dd52-48d5-a16b-3ff17149fba7 service=localauthority.v1.LocalAuthority subsystem_name=api
time="2025-08-26T13:35:23Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=TaintJWTAuthority request_id=5963ec52-dca2-4d90-89af-0d7b0a3ab0ec service=localauthority.v1.LocalAuthority subsystem_name=api
time="2025-08-26T13:35:23Z" level=error msg="JWT functionality is disabled" authorized_as=local authorized_via=transport method=RevokeJWTAuthority request_id=aad432cf-6f46-4cea-a983-89916b14f1be service=localauthority.v1.LocalAuthority subsystem_name=api
time="2025-08-26T13:35:23Z" level=error msg="JWT functionality is disabled" authorized_as=agent authorized_via=datastore caller_addr="10.244.0.1:61487" caller_id="spiffe://example.org/spire/agent/k8s_psat/example-cluster/eb94ef4d-e54b-4783-8165-ecb99d8806b7" method=NewJWTSVID request_id=b4697234-c82e-4d36-8789-1440235980a3 service=svid.v1.SVID subsystem_name=api
```

- i tested this with nested deployment and got mixed results. Servers where jwt is disabled cannot issue JWT but can validate JWT from other nested servers. I suppose this is working as expected?

- i did not find any federated trust domain integration test that uses JWT. I tried some things in ghosttunnel-federation test and in general came to opinion it wont be an issue. Trust bundles are exchanged and jwt validation was not disabled. (but feel free to suggest test cases to verify this properly)

Overall this fulfils our requirement:
- no jwt keys are managed
- jwt endpoints are disabled

Before i proceed with unit/integration tests and **documentation** please let me know if this is in general correct direction. 
I was a little lost with double abstraction levels - there is CA and CA manager, there is also CA rotator
Maybe you prefer disablement config in endpoints directly (however, CA manager would still need it to not manage JWT keys)

**Which issue this PR fixes**
<!-- optional. `fixes #<issue number>` format will close an issue when this PR is merged -->
fixes #4473

keeganwitt commented:
Capturing an idea @amartinezfayo had in the sync today: we might want to have the SPIRE server return a 501 UNIMPLEMENTED response when the OIDC provider tries to get JWKS, so that it can error out appropriately.

nikotih commented:
thank you @keeganwitt 

WIP thoughts

had a look at oidc provider code. found 2 usages:

a) talking to server directly via socket and calling getBundle and output mask JWT keys only
https://github.com/spiffe/spire/blob/main/support/oidc-discovery-provider/server_api.go#L112

Can set an exit branch for that if server has JWT disabled and get bundles is only asking for JWT keys
```
func (s *Service) GetBundle(ctx context.Context, req *bundlev1.GetBundleRequest) (*types.Bundle, error) {
	rpccontext.AddRPCAuditFields(ctx, logrus.Fields{telemetry.TrustDomainID: s.td.Name()})
	log := rpccontext.Logger(ctx)

	// if only JWKs are requested and we have JWT disabled
	if s.IsJWTDisabled() && s.req.OutputMask != nil && req.OutputMask.JwtAuthorities {
		return nil, api.MakeErr(log, codes.Unimplemented, "JWKs are not managed by this server", nil)
	}
```

Not the best solution but would work.
Another problem here is that bundles endpoint does not have CA or CA manager it uses datastore
So again might have to bring "disable_jwt" flag to endpoints or CA / CA manager

b) talking to agent via workload api and calling get bundle on local agent

this is more complicated, basically agent needs to know if jwt is disabled on server

can of course assume that if JwtAuthorities are empty in DB then its disabled? but that wont be the case for nested deployments or federated trust domains.

so agent needs to distinguish local server has jwt disabled but JwtAuthorities from nested / federated still in bundle

amartinezfayo commented:
Thank you @nikotih for working on this contribution.
I think that the approach in general looks good. I also explored if we could disable JWT support in the middleware, but I think that it will end up being a more complex implementation.

A couple of quick considerations:
- I think that we should return `codes.Unimplemented` as opposed to `codes.FailedPrecondition` when JWT-SVID support is disabled. I think that consumers receiving `codes.Unimplemented` can identify better that JWT-SVID support has been disabled in SPIRE.
- I would also suggest to name the setting `disable_jwt_svids`, to be more specific.
- We should log a message when the server has started with JWT SVID issuance disabled (probably at info level).

Have you looked at how this would work in the context of a nested SPIRE deployment? Particularly, if there are more than one level and one of the SPIRE servers in the middle has JWT-SVID disabled. It would be good to have this considered and clarity on the behavior.

nikotih commented:
Hi @amartinezfayo 

I've changed to codes.Unimplemented, renamed to disable_jwt_svids and added log line.

For nested setup i've modified original integration test and disabled jwt on intermediate B - it cannot issue JWTs but agent connected to it can still validate against other bundles

Apart from that i am more uncertain about parts of code that Keegan mentioned.
a) PruneBundle will need to know 0 jwtkeys might be ok - either flag that JWT is disabled have to be available there or it should not error on no keys in bundle
First will require datastore to know if jwt is disabled and propagate it all the way down (a lot of changes in interfaces), second might have side effects i am not aware of (0 keys in bundle is ok?)

b) Get Bundle issues mentioned above. If you want to simplify this in oidc provider to no JWT keys in bundle = jwt disabled that can work. Otherwise get bundle api becomes unpredictable and high cognitive complexity design:
- get all returns CAs and JWKs including from nested and federated
- get only JWKs will return unimplemented (despite bundle having JWKs from nested of federated)
- agent caching bundles will also need to know jwt is disabled - it gets all bundles from server (CA+JWT) so when e.g. oidc provider asks for bundles form agent - agent has no idea if its disabled on server

I would maybe think of some alternative way of propagating "JWT disabled" via grpc, some config sync between server<->agent, agent<->oidc provider and server<->oidc provider. This will keep bundle endpoints as is

If you have any ideas or suggestions please share

amartinezfayo commented on pkg/server/ca/ca.go:
```diff
@@ -391,6 +393,14 @@ func (ca *CA) signJWTSVID(jwtKey *JWTKey, claims map[string]any) (string, error)
+	return ca.c.DisableJWTSVIDs
+}
+
+func (ca *CA) SetDisableJWTSVIDs(disableJWTSVIDs bool) {
```
It seems like SetDisableJWTSVIDs is only used by tests?
If that's the case, I recommend removing this method entirely to prevent confusion and potential misuse in production code. The `DisableJWTSVIDs` setting should be configured at CA construction time only, and not provide the ability to be mutated at runtime.

Could we just have an attribute in the serviceTest struct indicating if JWTSVIDs are enabled or disabled? We can update that attribute as needed in the tests to exercise JWTSVIDs disabled scenarios.

nikotih replied:
from what i remember i had to add it due to `test/fakes/fakeserverca/serverca.go` overlay over CA. if there is no setter then that fake serverca has to store its own flag and ignore what is set in the actual CA

anyways i will have a look again tomorrow

nikotih commented on doc/spire_server.md:
```diff
@@ -68,6 +68,7 @@ This may be useful for templating configuration files, for example across differ
 | `default_jwt_svid_ttl`              | The default JWT-SVID TTL                                                                                                                                                                                                                        | 5m                                                             |
 | `experimental`                      | The experimental options that are subject to change or removal (see below)                                                                                                                                                                      |                                                                |
 | `federation`                        | Bundle endpoints configuration section used for [federation](#federation-configuration)                                                                                                                                                         |                                                                |
+| `disable_jwt_svids`                 | If true, disables JWT-SVID profile                                                                                                                                                                                                              | false                                                          |
```
Adjusted

amartinezfayo reviewed (approved):
Thank you again, @nikotih, for this valuable contribution.
The ability to disable JWT-SVIDs is something we’ve heard requests for from the community for quite some time. This will be included in the 1.14.0 release 🚀
